try_compile is stupid and can't handle alias targets lmao
---
 CMakeLists.txt | 45 +++++++++++++++++++++++++++++++++++++++------
 1 file changed, 39 insertions(+), 6 deletions(-)

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1015,7 +1015,24 @@
   if(USE_OPENSSL)
     list(APPEND CMAKE_REQUIRED_LIBRARIES OpenSSL::SSL OpenSSL::Crypto)
     if(HAVE_LIBZ)
-      list(APPEND CMAKE_REQUIRED_LIBRARIES ZLIB::ZLIB)
+      if(TARGET ZLIB::ZLIB)
+        get_target_property(_zlib_aliased ZLIB::ZLIB ALIASED_TARGET)
+        if(_zlib_aliased)
+          if(ZLIB_LIBRARIES)
+            list(APPEND CMAKE_REQUIRED_LIBRARIES ${ZLIB_LIBRARIES})
+          endif()
+          if(ZLIB_INCLUDE_DIRS)
+            list(APPEND CMAKE_REQUIRED_INCLUDES ${ZLIB_INCLUDE_DIRS})
+          elseif(ZLIB_INCLUDE_DIR)
+            list(APPEND CMAKE_REQUIRED_INCLUDES ${ZLIB_INCLUDE_DIR})
+          endif()
+        else()
+          list(APPEND CMAKE_REQUIRED_LIBRARIES ZLIB::ZLIB)
+        endif()
+        unset(_zlib_aliased)
+      elseif(ZLIB_LIBRARIES)
+        list(APPEND CMAKE_REQUIRED_LIBRARIES ${ZLIB_LIBRARIES})
+      endif()
     endif()
     if(WIN32)
       list(APPEND CMAKE_REQUIRED_LIBRARIES "bcrypt")  # for OpenSSL/LibreSSL BCryptGenRandom()
@@ -1025,7 +1043,25 @@
     list(APPEND CMAKE_REQUIRED_LIBRARIES CURL::wolfssl)
     list(APPEND CMAKE_REQUIRED_DEFINITIONS "-DWOLFSSL_OPTIONS_IGNORE_SYS")
     if(HAVE_LIBZ)
-      list(APPEND CMAKE_REQUIRED_LIBRARIES ZLIB::ZLIB)  # Public wolfSSL headers also require zlib headers
+      # Public wolfSSL headers also require zlib headers
+      if(TARGET ZLIB::ZLIB)
+        get_target_property(_zlib_aliased ZLIB::ZLIB ALIASED_TARGET)
+        if(_zlib_aliased)
+          if(ZLIB_LIBRARIES)
+            list(APPEND CMAKE_REQUIRED_LIBRARIES ${ZLIB_LIBRARIES})
+          endif()
+          if(ZLIB_INCLUDE_DIRS)
+            list(APPEND CMAKE_REQUIRED_INCLUDES ${ZLIB_INCLUDE_DIRS})
+          elseif(ZLIB_INCLUDE_DIR)
+            list(APPEND CMAKE_REQUIRED_INCLUDES ${ZLIB_INCLUDE_DIR})
+          endif()
+        else()
+          list(APPEND CMAKE_REQUIRED_LIBRARIES ZLIB::ZLIB)
+        endif()
+        unset(_zlib_aliased)
+      elseif(ZLIB_LIBRARIES)
+        list(APPEND CMAKE_REQUIRED_LIBRARIES ${ZLIB_LIBRARIES})
+      endif()
     endif()
     list(APPEND CMAKE_REQUIRED_DEFINITIONS "-DHAVE_UINTPTR_T")  # to pull in stdint.h (as of wolfSSL 5.5.4)
   endif()
